File: arduino_bridge.js
// Arduino Serial Bridge - Node.js version
const { SerialPort } = require('serialport');
const { ReadlineParser } = require('@serialport/parser-readline');
const https = require('https');
const http = require('http');

// Configuration
const BACKEND_URL = 'http://localhost:3001/api/arduino/number';
const BAUD_RATE = 9600;

// Find Arduino port
async function findArduinoPort() {
    try {
        const ports = await SerialPort.list();
        console.log('🔍 Available ports:');
        ports.forEach(port => {
            console.log(`  - ${port.path}: ${port.manufacturer || 'Unknown'}`);
        });
        
        // Try to find Arduino - prioritize COM14 per user
        let arduinoPort = ports.find(port => port.path === 'COM14');
        
        if (!arduinoPort) {
            // Fallback to other detection methods
            arduinoPort = ports.find(port => 
                port.manufacturer && (
                    port.manufacturer.includes('Arduino') ||
                    port.manufacturer.includes('FTDI') ||
                    port.manufacturer.includes('CH340')
                )
            );
        }
        
        if (arduinoPort) {
            console.log(`✅ Found Arduino on: ${arduinoPort.path}`);
            return arduinoPort.path;
        } else {
            console.log('❌ Arduino not found automatically');
            console.log('💡 Trying COM14 (user specified)...');
            return 'COM14'; // Use COM14 as specified by user
        }
    } catch (error) {
        console.error('Error listing ports:', error.message);
        return 'COM3'; // Default fallback
    }
}

// Send to backend using native http
function sendToBackend(number) {
    return new Promise((resolve) => {
        const data = JSON.stringify({ number: number.toString() });
        
        const options = {
            hostname: 'localhost',
            port: 3001,
            path: '/api/arduino/number',
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(data)
            }
        };
        
        const req = http.request(options, (res) => {
            let responseData = '';
            
            res.on('data', (chunk) => {
                responseData += chunk;
            });
            
            res.on('end', () => {
                if (res.statusCode === 200) {
                    console.log(`✅ Sent to backend: ${number}`);
                    resolve(true);
                } else {
                    console.log(`❌ Backend error: ${res.statusCode}`);
                    resolve(false);
                }
            });
        });
        
        req.on('error', (error) => {
            console.log(`❌ Failed to send to backend: ${error.message}`);
            resolve(false);
        });
        
        req.write(data);
        req.end();
    });
}

// Main function
async function main() {
    console.log('🔌 Arduino Serial Bridge Starting...');
    console.log(`🎯 Backend URL: ${BACKEND_URL}`);
    
    // Find Arduino port
    const portPath = await findArduinoPort();
    
    try {
        // Connect to Arduino
        const port = new SerialPort({
            path: portPath,
            baudRate: BAUD_RATE
        });
        
        const parser = port.pipe(new ReadlineParser({ delimiter: '\n' }));
        
        port.on('open', () => {
            console.log(`📡 Connected to Arduino on ${portPath}`);
            console.log('📱 Send numbers from your Bluetooth app!');
        });
        
        parser.on('data', (data) => {
            const line = data.toString().trim();
            console.log(`📡 Arduino: ${line}`);
            
            // Extract numbers from the line
            const numbers = line.match(/\d+/g);
            
            if (numbers && numbers.length > 0) {
                const number = numbers[0]; // Take first number found
                console.log(`📱 Detected number: ${number}`);
                sendToBackend(number);
            }
        });
        
        port.on('error', (error) => {
            console.error(`❌ Serial port error: ${error.message}`);
            console.log('\n💡 Troubleshooting:');
            console.log('1. Check if Arduino is connected via USB');
            console.log('2. Close Arduino IDE Serial Monitor if open');
            console.log('3. Try different COM port (COM3, COM4, COM5)');
            console.log('4. Check Device Manager > Ports (COM & LPT)');
        });
        
    } catch (error) {
        console.error(`❌ Failed to connect: ${error.message}`);
        console.log('\n💡 Troubleshooting:');
        console.log('1. Install: npm install serialport');
        console.log('2. Check Arduino connection');
        console.log('3. Close other programs using the port');
    }
}

// Handle Ctrl+C
process.on('SIGINT', () => {
    console.log('\n🛑 Stopping Arduino bridge...');
    process.exit(0);
});

// Start the bridge
main().catch(console.error);

File: arduino_bridge.py
#!/usr/bin/env python3
"""
Arduino Serial Bridge - Reads from Arduino and sends to backend
"""

import serial
import requests
import time
import json
import sys

# Configuration
BACKEND_URL = "http://localhost:3001/api/arduino/number"
BACKEND_RAW_URL = "http://localhost:3001/api/arduino/data"
ARDUINO_PORT = "COM14"  # Default to your Arduino port (COM14)
BAUD_RATE = 9600

def find_arduino_port():
    """Try to find Arduino on common ports"""
    # Force strict port usage
    try:
        ser = serial.Serial(ARDUINO_PORT, BAUD_RATE, timeout=1)
        print(f"✅ Found Arduino on {ARDUINO_PORT}")
        return ser
    except Exception as e:
        print(f"❌ Could not open {ARDUINO_PORT}: {e}")
        print("➡️  Please verify in Device Manager that the Arduino is on COM14 and close any Serial Monitor.")
        return None

def send_to_backend(number):
    """Send number to backend API"""
    try:
        data = {"number": str(number)}
        response = requests.post(BACKEND_URL, json=data, timeout=5)
        if response.status_code == 200:
            print(f"✅ Sent to backend: {number}")
            return True
        else:
            print(f"❌ Backend error: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Failed to send to backend: {e}")
        return False

def send_raw_to_backend(raw_line):
    """Send raw data line to backend (used for NO_MATCH and diagnostics)"""
    try:
        data = {"data": str(raw_line)}
        response = requests.post(BACKEND_RAW_URL, json=data, timeout=5)
        if response.status_code == 200:
            print(f"✅ Sent raw to backend: {raw_line}")
            return True
        else:
            print(f"❌ Backend raw error: {response.status_code}")
            return False
    except Exception as e:
        print(f"❌ Failed to send raw to backend: {e}")
        return False

def main():
    print("🔌 Arduino Serial Bridge Starting...")
    print(f"🎯 Backend URL: {BACKEND_URL}")
    
    # Find Arduino
    arduino = find_arduino_port()
    if not arduino:
        print("\n💡 To find your Arduino port:")
        print("1. Open Device Manager")
        print("2. Look under 'Ports (COM & LPT)'")
        print("3. Find 'Arduino Uno (COMx)' or similar")
        print("4. Update ARDUINO_PORT in this script")
        return
    
    print(f"📡 Listening for fingerprint data on {arduino.port}...")
    print("👉 Place finger on sensor. Expect lines like 'VOTER: <id>' or any numeric template ID.")

    # Clear any stale bytes to avoid partial/garbled frames
    try:
        arduino.reset_input_buffer()
    except Exception:
        pass

    try:
        while True:
            if arduino.in_waiting > 0:
                # Read a line as bytes and decode robustly
                raw = arduino.readline()
                if not raw:
                    continue
                line = raw.decode('utf-8', errors='ignore').strip()

                if line:
                    print(f"📡 Arduino: {line}")

                    # If Arduino reports NO_MATCH, notify backend to show error immediately
                    if 'NO_MATCH' in line.upper():
                        send_raw_to_backend(line)
                        continue

                    # Extract numbers (fingerprint template ID) from the line
                    import re
                    numbers = re.findall(r'\d+', line)

                    for number in numbers:
                        if len(number) > 0:  # Any number length
                            print(f"📱 Detected number: {number}")
                            send_to_backend(number)
                            break  # Process only first number found
            
            time.sleep(0.1)  # Small delay
            
    except KeyboardInterrupt:
        print("\n🛑 Stopping Arduino bridge...")
    except Exception as e:
        # Log and keep running instead of crashing on transient errors
        print(f"❌ Error: {e}")
    finally:
        if arduino:
            arduino.close()
        print("👋 Arduino bridge stopped")

if __name__ == "__main__":
    main()